### ISSUE TITLE
Oauth flow: missing state, response type, does not grab code

### ISSUE THREAD

**Issue by synotna** [frigus02/RESTer#42]
Version: 3.2.1
Browser: Firefox 57.0b12 (64-bit)

I'm trying to use a normal OAuth 2 flow but
* rester does not add a state parameter by itself
** worked around by adding to the auth url myself (?state=a)
* rester does not add a response type parameter by itself
** worked around by adding to the auth url myself (?response_type=code)
* rester does not grab the code to to finish the flow
** can't work around

**Comment by frigus02** [comment]
Hi @synotna, thanks for the bug report. I cannot fully reproduce the issue, but I noticed some bugs. Can you help me to track down the root cause?

1.  RESTer does not send the `state` parameter. I did not implement this, because it is optional in the spec and I didn't think it would be useful. I'm really interested in your use case for it, though. Can you describe what would like to do?

2.  Adding a parameter to the authentication endpoint URL like `url?state=...` does break the OAuth flow. RESTer will always append the parameters after another question mark, so the URL will look like this:

    ```
    http://url/auth?state=123?response_type=code&...
    ```

    This is a bug, which I will fix in the next update.

    I assume this is why you don't see the response_type parameter appended. If you try it with an authentication endpoint URL without a query string, you should see all necessary parameters being appended.

3.  RESTer is not able to grab the code from the redirect URL if this URL fails to load. This is new to me and I need to investigate it. After a short research it seems to be a restriction of the browser extension model, though.

    So in short: The redirect URL needs to be a valid and reachable URL. I'm able to use e.g. `https://github.com`, when I configured this in my identity provider.

**Comment by frigus02** [comment]
I just uploaded version 3.3.0, which fixes the bug about incorrectly appending the query string. I'm assuming this fixes your issues, so I will close this now. Feel free to reopen it, if your issue is something different.
